pages/api/progress.py

```diff
@@ -253,21 +270,77 @@ def post(self, request, *args, **kwargs):
-            progress.save()
+        if live_event.scheduled_at >= datetime_or_now():
             return api_response.Response(
-                {'detail': 'Attendance marked successfully'},
-                status=status.HTTP_200_OK)
+                {'detail': 'Can not mark attendance for an event at a future date'},
+                status=status.HTTP_400_BAD_REQUEST)
+
+        extra = json.loads(progress.extra) if progress.extra else {}
+
+        if get_extra(progress, "live_event_rank") == event_rank and \
+            progress.viewing_duration >= progress.step.min_viewing_duration:
+
+            return api_response.Response({'detail': 'Attendance already marked'}, 
+                                         status=status.HTTP_400_BAD_REQUEST)
+
+        if progress.viewing_duration < progress.step.min_viewing_duration:
+            progress.viewing_duration = progress.step.min_viewing_duration
+
+        extra["live_event_rank"] = event_rank
+        progress.extra = json.dumps(extra)
+        progress.save()
         return api_response.Response(
-            {'detail': 'Attendance not marked'},
-            status=status.HTTP_400_BAD_REQUEST)
+            {'detail': 'Attendance marked successfully'}, status=status.HTTP_201_CREATED)
+    
+    def delete(self, request, *args, **kwargs):
+        '''
+        Resets Live Event Attendance
+        '''
+        # Maybe redundant because we're already resetting EnumeratedProgress?
+        event_rank = kwargs.get(self.event_rank_kwarg)
+        progress = self.get_object()
+
+        extra = json.loads(progress.extra) if progress.extra else {}
+
+        if get_extra(progress, "live_event_rank") == event_rank:
+            extra["live_event_rank"] = ""
+            progress.extra = json.dumps(extra)
+            progress.viewing_duration = datetime.timedelta(seconds=0)
+            progress.save()
+            return api_response.Response(status=status.HTTP_204_NO_CONTENT)
+
+        return api_response.Response(status=status.HTTP_400_BAD_REQUEST)
+
+
+class LiveEventAttendeesAPIView(SequenceMixin, ListAPIView):
+    serializer_class = LiveEventAttendeesSerializer
+
+    rank_url_kwarg = 'rank'
+    event_rank_kwarg = 'event_rank'
+
+    def get_queryset(self):
+        element_rank = self.kwargs.get(self.rank_url_kwarg)
+        event_rank = self.kwargs.get(self.event_rank_kwarg)
+
+        element = get_object_or_404(
+            EnumeratedElements, sequence=self.sequence, rank=element_rank)
+
+        queryset = EnumeratedProgress.objects.filter(
+            step__content=element.content,
+            viewing_duration__gte=element.min_viewing_duration)
+    
+        progress_ids = [progress.id for progress in queryset if 
+                get_extra(progress, "live_event_rank") == event_rank]
+        queryset = queryset.filter(id__in=progress_ids)
+        return queryset
+
+    def get(self, request, *args, **kwargs):
```

Comment:
This method does not seem necessary as `LiveEventAttendeesAPIView` inherits from `ListAPIView`.